This notebook is used for exploring the data to process.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('../data/raw/historical_stock.csv', index_col=0, header=[0,1])

df = pd.DataFrame(data)

df.head(10)

Ticker           SAN                                              IBE.MC  \
Price           Open      High       Low     Close      Volume      Open   
Date                                                                       
2022-01-03  2.837175  2.880033  2.837175  2.880033   5420200.0  8.406721   
2022-01-04  2.931462  2.965748  2.922890  2.940033   6220000.0  8.503909   
2022-01-05  2.965748  2.982891  2.897176  2.914319   4023800.0  8.406720   
2022-01-06  2.991462  3.042891  2.957176  3.034320   5543000.0  8.281186   
2022-01-07  3.008606  3.068606  3.000034  3.068606   7203300.0  8.269037   
2022-01-10  3.077178  3.077178  3.017177  3.077178  11876700.0  8.354896   
2022-01-11  3.034320  3.111464  3.008605  3.094321   8275900.0  8.227247   
2022-01-12  3.060035  3.094321  3.051463  3.085749   5875000.0  8.276660   
2022-01-13  3.094321  3.120035  3.068606  3.085749   9044000.0  8.210775   
2022-01-14  3.077178  3.162893  3.068606  3.162893   8919500.0  8.194304   

Ticker                                                ...     ITX.MC  \
Price           High       Low     Close      Volume  ...       Open   
Date                                                  ...              
2022-01-03  8.507958  8.406721  8.459364   8281229.0  ...  24.317524   
2022-01-04  8.520108  8.366227  8.410771   9869738.0  ...  24.589136   
2022-01-05  8.426967  8.333829  8.354076  12421625.0  ...  24.190211   
2022-01-06  8.321680  8.244740  8.301433  12555060.0  ...  23.765822   
2022-01-07  8.354076  8.204246  8.305483  12965678.0  ...  23.927089   
2022-01-10  8.354896  8.149832  8.251952  10722150.0  ...  23.630018   
2022-01-11  8.284896  8.207482  8.256071  14101927.0  ...  23.774308   
2022-01-12  8.280777  8.192657  8.227246  18614818.0  ...  23.842211   
2022-01-13  8.359013  8.197599  8.243717  10892929.0  ...  23.638503   
2022-01-14  8.205834  8.059243  8.179481  12449056.0  ...  23.528161   

Ticker                                                      AIR.PA  \
Price            High        Low      Close     Volume        Open   
Date                                                                 
2022-01-03  24.665523  24.224159  24.326012  1590145.0  104.220618   
2022-01-04  24.589136  24.113820  24.198698  1933130.0  106.954593   
2022-01-05  24.376942  24.071383  24.156260  1809780.0  109.468351   
2022-01-06  24.079871  23.596066  23.994993  2829840.0  108.440832   
2022-01-07  24.037430  23.383870  23.477236  2276001.0  108.257335   
2022-01-10  23.799774  23.409335  23.494213  2585589.0  108.257341   
2022-01-11  23.969526  23.655479  23.714893  1888616.0  108.807808   
2022-01-12  23.876162  23.672455  23.672455  3272888.0  108.917886   
2022-01-13  23.884648  23.426308  23.663965  2535858.0  108.257337   
2022-01-14  23.638503  23.400844  23.570601  1782131.0  106.679354   

Ticker                                                     
Price             High         Low       Close     Volume  
Date                                                       
2022-01-03  106.660998  103.871995  106.550903  1320377.0  
2022-01-04  109.963779  106.422479  109.064690  1588549.0  
2022-01-05  111.101384  107.523386  110.459183  1806200.0  
2022-01-06  110.183960  107.688535  108.844505  1441589.0  
2022-01-07  108.826147  106.734389  107.982101  1244642.0  
2022-01-10  109.376613  106.954585  108.183945  1099153.0  
2022-01-11  109.266526  107.468354  108.642670  1060555.0  
2022-01-12  109.835321  107.872010  108.881187  1288629.0  
2022-01-13  108.679356  106.605951  107.615135  1278350.0  
2022-01-14  108.312387  106.587611  107.651833  1051871.0  

[10 rows x 25 columns]

In [3]:
df.info() 
df.isnull().sum()

<class 'pandas.DataFrame'>
Index: 775 entries, 2022-01-03 to 2024-12-31
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   (SAN, Open)       753 non-null    float64
 1   (SAN, High)       753 non-null    float64
 2   (SAN, Low)        753 non-null    float64
 3   (SAN, Close)      753 non-null    float64
 4   (SAN, Volume)     753 non-null    float64
 5   (IBE.MC, Open)    768 non-null    float64
 6   (IBE.MC, High)    768 non-null    float64
 7   (IBE.MC, Low)     768 non-null    float64
 8   (IBE.MC, Close)   768 non-null    float64
 9   (IBE.MC, Volume)  768 non-null    float64
 10  (SIE.DE, Open)    766 non-null    float64
 11  (SIE.DE, High)    766 non-null    float64
 12  (SIE.DE, Low)     766 non-null    float64
 13  (SIE.DE, Close)   766 non-null    float64
 14  (SIE.DE, Volume)  766 non-null    float64
 15  (ITX.MC, Open)    768 non-null    float64
 16  (ITX.MC, High)    768 non-null    float64
 1

Ticker  Price 
SAN     Open      22
        High      22
        Low       22
        Close     22
        Volume    22
IBE.MC  Open       7
        High       7
        Low        7
        Close      7
        Volume     7
SIE.DE  Open       9
        High       9
        Low        9
        Close      9
        Volume     9
ITX.MC  Open       7
        High       7
        Low        7
        Close      7
        Volume     7
AIR.PA  Open       7
        High       7
        Low        7
        Close      7
        Volume     7
dtype: int64

The *historical_stock.csv* file has 775 entries, from **2022-01-01** to **2025-01-01**. It is easyly observable that at least 7 days are null in every column, those days need to be assessed, as well as some other faulty registries clearly shown in the information above. 

In [4]:
null_dates = df.isnull()

null_dates = null_dates[null_dates.any(axis=1)]
null_dates = null_dates.index.tolist()
print(null_dates)
print(f"Number of null dates: {len(null_dates)}")
print(f"Percentage of null dates: {len(null_dates)/len(df)*100:.2f}%")

['2022-01-17', '2022-02-21', '2022-04-18', '2022-05-30', '2022-06-20', '2022-07-04', '2022-09-05', '2022-11-24', '2023-01-02', '2023-01-16', '2023-02-20', '2023-04-10', '2023-05-01', '2023-05-29', '2023-06-19', '2023-07-04', '2023-09-04', '2023-11-23', '2023-12-26', '2024-01-15', '2024-02-19', '2024-04-01', '2024-05-01', '2024-05-27', '2024-06-19', '2024-07-04', '2024-09-02', '2024-11-28', '2024-12-24', '2024-12-26', '2024-12-31']
Number of null dates: 31
Percentage of null dates: 4.00%


Being the percentage of null dates as low as $4\%$, these registries can be deleted without compromising the integrity of this study. 

In [5]:
df.dropna(inplace=True)